File: 프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.cpp
#include <bits/stdc++.h>

using namespace std;

vector<long long> solution(vector<long long> numbers) {
    vector<long long> answer;
    
    for (long long number : numbers) {
        
        // 짝수인 경우
        if (number % 2 == 0) {
            answer.push_back(number + 1);
        }
        
        // 홀수인 경우
        else {
            string binary = bitset<64>(number).to_string();
            int index = binary.rfind("0");
            answer.push_back(stoll(binary.replace(index, 2, "10"), nullptr, 2));
        }
    }
    
    return answer;
}

File: 프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.java
class Solution {
    
    public long[] solution(long[] numbers) {
        
        long[] answer = new long[numbers.length];
        for (int i = 0; i < numbers.length; i++) {
            
            // 짝수인 경우
            if (numbers[i] % 2 == 0) {
                // answer[i] = numbers[i] | 1;
                answer[i] = numbers[i] + 1;
            }
            
            // 홀수인 경우
            else {
                String number = Long.toBinaryString(numbers[i]);
                int index = number.lastIndexOf("0");    
                
                // 모든 바이너리 숫자가 1로 구성된 경우
                if (index == -1) {
                    number = "10" + number.substring(1);
                }
                // 바이너리 숫자 중 0이 존재하는 경우
                else {
                    number = number.substring(0, index) + "10" + number.substring(index+2);
                }
                
                answer[i] = Long.parseLong(number, 2);        
            }
        }
        
        return answer;
    }
}